[clang-tidy] Improved modernize-use-using by fixing a false-negative

felix642 opened the pull request:
The check needs a parent decl to match but if the typedef is in a function, the parent is a declStmt which is not a decl by itself. Improved the matcher to match on either a decl or a declstmt and extract the decl from the stmt in the latter case.

fixes #72179

CC: @FabianWolff @PiotrZSL @njames93

felix642 committed:
[clang-tidy] Improved modernize-use-using by fixing a false-negative

The check needs a parent decl to match but if the typedef is in a function,
the parent is a declStmt which is not a decl by itself.
Improved the matcher to match on either a decl or a declstmt and extract
the decl from the stmt in the latter case.

fixes #72179

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -183,6 +183,9 @@ Changes in existing checks
   whitespace when deleting the ``virtual`` keyword.
 
+- Improved :doc:`modernize-use-using <clang-tidy/checks/modernize/use-using>`
+  check by fixing false-negative in functions.
```
maybe: `by adding support for detection of typedefs declared on function level.`

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/modernize/use-using.cpp:
```diff
@@ -342,3 +342,21 @@ typedef int InExternCPP;
+    // CHECK-MESSAGES: :[[@LINE-1]]:5: warning: use 'using' instead of 'typedef' [modernize-use-using]
+    // CHECK-FIXES: using c = struct { int a; union { int b; }; };
+  }
+}
```
add test with lambda, template function, normal function in template class.

PiotrZSL reviewed (approved):
Overall fine.

felix642 commented:
Thank you for the quick review @PiotrZSL. I've implemented the proposed changes and rebased on top of main. We can wait a few days to see if someone else has some comments to add and then we should be ready to merge

felix642 commented:
@PiotrZSL I've squashed my changed. When you'll have time, I think we can merge this PR.
